index.d.ts

```diff
@@ -0,0 +1,355 @@
+    introspectionEndpointAuthMethods?: AuthMethod[];
+    logoutSource?: (ctx: Koa.Context, form: string) => void;
+    pairwiseIdentifier?: (accountId: string, client: object) => string;
+    postLogoutRedirectUri?: (ctx: Koa.Context) => string;
+    prompts?: string[];
+    refreshTokenRotation?: "rotateAndConsume" | "none";
+    renderError?: (ctx: Koa.Context, out: object, error: Error) => void;
+    responseTypes?: string[];
+    revocationEndpointAuthMethods?: AuthMethod[];
+    routes?: IRoutes;
+    scopes?: string[];
+    subjectTypes?: SubjectType;
+    tokenEndpointAuthMethods?: AuthMethod[];
+    ttl?: {
+      AccessToken?: number;
+      AuthorizationCode?: number;
+      ClientCredentials?: number;
+      DeviceCode?: number;
+      IdToken?: number;
+      RefreshToken?: number;
+    };
+    uniqueness?: (ctx: Koa.Context, jti: string, expiresAt: number) => boolean;
+    userCodeConfirmSource?: (
+      ctx: Koa.Context,
+      form: string,
+      client: object,
+      deviceInfo: object
+    ) => void;
+    userCodeInputSource?: (ctx: Koa.Context, form: string, out: object, err: Error) => void;
+    whitelistedJWA?: {
+      authorizationEncryptionAlgValues?: EncryptionAlgValue[];
+      authorizationEncryptionEncValues?: EncryptionEncValue[];
+      authorizationSigningAlgValues?: SigningAlgValue[];
+      idTokenEncryptionAlgValues?: EncryptionAlgValue[];
+      idTokenEncryptionEncValues?: EncryptionEncValue[];
+      idTokenSigningAlgValues?: SigningAlgValue[];
+      introspectionEncryptionAlgValues?: EncryptionAlgValue[];
+      introspectionEncryptionEncValues?: EncryptionEncValue[];
+      introspectionEndpointAuthSigningAlgValues?: SigningAlgValue[];
+      introspectionSigningAlgValues?: SigningAlgValue[];
+      requestObjectEncryptionAlgValues?: EncryptionAlgValue[];
+      requestObjectEncryptionEncValues?: EncryptionEncValue[];
+      requestObjectSigningAlgValues?: SigningAlgValue[];
+      revocationEndpointAuthSigningAlgValues?: SigningAlgValue[];
+      tokenEndpointAuthSigningAlgValues?: SigningAlgValue[];
+      userinfoEncryptionAlgValues?: EncryptionAlgValue[];
+      userinfoEncryptionEncValues?: EncryptionEncValue[];
+      userinfoSigningAlgValues?: SigningAlgValue[];
+    };
+  }
+
+  class Provider extends EventEmitter {
+    app: Koa;
+
+    domain: {};
+
+    proxy: boolean;
+
+    constructor(url: string, config?: IConfiguration);
+
+    initialize(config: {
+      adapter?: new (name: string) => IAdapter;
+      clients?: IClient[];
+      keystore?: unknown;
+    }): Promise<this>;
+
+    interactionDetails(ctx: IncomingMessage): Promise<ISession>;
+
+    setProviderSession(req: IncomingMessage, res: ServerResponse, {}): Promise<ISession>;
+
+    interactionFinished(req: IncomingMessage, res: ServerResponse, {}): Promise<void>;
+
+    callback: Handler;
+    listen(port: string | number): void;
+
+    Client: IClient;
```

Comment:
i'd say minimal types don't need to type classes that are used for advanced customization, so please remove Client from here, same for `session`